File: java-packages/src/main/java/com/rmaciak/order/domain/Order.java
package com.rmaciak.order.domain;

import java.math.BigDecimal;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

final class Order {

    private final UUID id;
    private final BigDecimal quota;
    private final Set<OrderItem> items;

    Order(UUID id, BigDecimal quota, Set<OrderItem> items) {
        this.id = id;
        this.quota = quota;
        this.items = items;
    }

    static Order createOrder() {
        return new Order(
                UUID.randomUUID(),
                BigDecimal.ZERO,
                Set.of()
        );
    }

    Order addItem(OrderItem orderItem) {
        Set<OrderItem> newItems = new HashSet<>(this.items);
        newItems.add(orderItem);

        return new Order(
                id,
                this.quota.add(orderItem.itemQuota()),
                newItems
        );
    }

    UUID orderId() {
        return id;
    }

    BigDecimal orderQuota() {
        return quota;
    }

    Set<OrderItem> orderItems() {
        return items;
    }
}

File: kotlin-with-submodules/order-domain/src/main/kotlin/com/rmaciak/order/Order.kt
package com.rmaciak.order

import java.math.BigDecimal
import java.math.BigDecimal.ZERO
import java.util.UUID

internal data class Order(
        private val id: UUID,
        private val quota: BigDecimal,
        private val items: Set<OrderItem>
) {

    internal fun addItem(item: OrderItem) =
            this.copy(
                    quota = this.quota.add(item.itemQuota()),
                    items = items + item
            )

    companion object {
        fun create() = Order(UUID.randomUUID(), ZERO, emptySet())
    }

    internal fun toEntity() =
            OrderEntity(
                    id = this.id,
                    quota = this.quota,
                    items = this.items.map { it.toEntity() }.toSet()
            )

    internal fun toDto() =
            OrderDto(
                    id = this.id,
                    quota = this.quota,
                    items = this.items.map { it.toDto() }.toSet()
            )
}